x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/bulk/BulkShardOperationsTests.java

```diff
@@ -78,4 +87,114 @@ public void testPrimaryTermFromFollower() throws IOException {
         closeShards(followerPrimary);
     }
 
+    public void testPrimaryResultWaitForGlobalCheckpoint() throws Exception {
+        final Settings settings = Settings.builder().put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true).build();
+        final IndexShard shard = newStartedShard(false, settings, new FollowingEngineFactory());
+        int numOps = between(1, 100);
+        for (int i = 0; i < numOps; i++) {
+            final String id = Integer.toString(i);
+            final Translog.Operation op;
+            if (randomBoolean()) {
+                op = new Translog.Index("_doc", id, i, primaryTerm, 0, SOURCE, null, -1);
+            } else if (randomBoolean()) {
+                shard.advanceMaxSeqNoOfUpdatesOrDeletes(i);
+                op = new Translog.Delete("_doc", id, new Term("_id", Uid.encodeId(id)), i, primaryTerm, 0);
+            } else {
+                op = new Translog.NoOp(i, primaryTerm, "test");
+            }
+            shard.applyTranslogOperation(op, Engine.Operation.Origin.REPLICA);
+        }
+        BulkShardOperationsRequest request = new BulkShardOperationsRequest();
+        {
+            PlainActionFuture<BulkShardOperationsResponse> listener = new PlainActionFuture<>();
+            CcrWritePrimaryResult primaryResult = new CcrWritePrimaryResult(request, null, shard, -2, logger);
+            primaryResult.respond(listener);
+            assertThat("should return intermediately if waiting_global_checkpoint is not specified", listener.isDone(), equalTo(true));
+            assertThat(listener.get().getMaxSeqNo(), equalTo(shard.seqNoStats().getMaxSeqNo()));
+        }
+        {
+            PlainActionFuture<BulkShardOperationsResponse> listener = new PlainActionFuture<>();
+            long waitingForGlobalCheckpoint = randomLongBetween(shard.getGlobalCheckpoint() + 1, shard.getLocalCheckpoint());
+            CcrWritePrimaryResult primaryResult = new CcrWritePrimaryResult(request, null, shard, waitingForGlobalCheckpoint, logger);
+            primaryResult.respond(listener);
+            assertThat(listener.isDone(), equalTo(false));
+            expectThrows(ElasticsearchTimeoutException.class, () -> listener.actionGet(TimeValue.timeValueMillis(1)));
+
+            shard.updateGlobalCheckpointOnReplica(randomLongBetween(shard.getGlobalCheckpoint(), waitingForGlobalCheckpoint - 1), "test");
+            expectThrows(ElasticsearchTimeoutException.class, () -> listener.actionGet(TimeValue.timeValueMillis(1)));
+
+            shard.updateGlobalCheckpointOnReplica(randomLongBetween(waitingForGlobalCheckpoint, shard.getLocalCheckpoint()), "test");
+            assertThat(listener.actionGet(TimeValue.timeValueSeconds(5)).getMaxSeqNo(), equalTo(shard.seqNoStats().getMaxSeqNo()));
```

Comment:
can we make this just get()? I'm not so comfortable with 5s (it's short) but also we typically let the suite time out so we can get a thread dump (although I suspect it won't be that helpful here, it might)